src/corona_hakab_model/affinity_matrix.py

```diff
@@ -212,17 +88,169 @@ def normalize(self):
+
+            # temp holder for used agents, so that the same connection wont be made twice
+            temp_agents_holder = []
+
+            # creating all of first's connections
+            for _ in range(remaining_contacts[current_agent]):
+                if connections_sum <= 1 or available_agents_len <= 1:
+                    break
+
+                # choosing 2nd agent for the connection
+                # todo make sure the % operator doesn't harm the randomness too much
+                roll_index = rolls[roll_cnt] % available_agents_len
+                roll_cnt += 1
+                second_agent = available_agents[roll_index]
+
+                # adding the newly made connection to the to-be-added connections structure
+                connections_tuples[0, connections_cnt] = current_agent.index
+                connections_tuples[1, connections_cnt] = second_agent.index
+                connections_cnt += 1
+                connections_tuples[0, connections_cnt] = second_agent.index
+                connections_tuples[1, connections_cnt] = current_agent.index
+                connections_cnt += 1
+
+                # note that there is no reason to update firs's remaining connections for efficiency
+                remaining_contacts[second_agent] -= 1
+                connections_sum -= 2
+
+                if remaining_contacts[second_agent] <= 0:
+                    del (available_agents[roll_index])
+                    available_agents_len -= 1
+                else:
+                    temp_agents_holder.append(available_agents.pop(roll_index))
+                    available_agents_len -= 1
+            # returning all the agents back from the temp place holder
+            available_agents_len += len(temp_agents_holder)
+            available_agents.extend(temp_agents_holder)
+
+        # filling the matrix. sometimes there still remains 1 un-filled connection, and it is left as 0,0 connection, so reset 0,0
+        matrix[connections_tuples[0], connections_tuples[1]] = connection_strength
+        matrix[0, 0] = 0
+
+        return matrix
+
+    def circular_matrix_generation(self, agents, circle_size_probability: rv_discrete, connection_strength):
+        """
+        this method will create a matrix of circular connections given a list of agents, an rv_discrete representing the size of the circle probabilty, and the connection strength
+        :param agents: list of all agents
+        :param circle_size_probability: representing the size probability
+        :param connection_strength: the wanted connection strength
+        :return: lil_matrix with the wanted connections
+        """
+        my_agents_length = len(agents)
+        matrix = m_type((self.size, self.size), dtype=np.float32)
+
+        # pre-rolling all the sized of the circles for efficiency causes.
+        circles_size_rolls = circle_size_probability.rvs(
+            size=math.ceil(my_agents_length / circle_size_probability.mean()))
```

Comment:
would should take at least one SD more the we would expect, random number generation is faster in bulk. (circle_size_probability also has a variance).

Reply:
also this should be an iterator

Reply:
If you only roll `my_agents_length / circle_size_probability.mean()` values, you will have roll more values half the time, if you do `my_agents_length / circle_size_probability.mean() + 10`, then that's less likely. Of course with some match and the SD of the distribution we can get a more accurate additional rolls that will only require re-rolling by some small probability.